sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MetricStorageRegistry.java

```diff
@@ -5,71 +5,88 @@
-   * equal with the current metric, otherwise throws an exception.
+   * Registers the metric {@code newStorage} to this registry. If a metric with compatible identity
+   * was previously registered, returns the previously registered storage. If a metric with the same
+   * name (case-insensitive) but incompatible {@link MetricDescriptor} was previously registered,
+   * logs a diagnostic warning and returns the {@code newStorage}.
    *
-   * @param storage the metric storage to use or discard.
-   * @return the given metric storage if no metric with same name already registered, otherwise the
-   *     previous registered instrument.
-   * @throws IllegalArgumentException if instrument cannot be registered.
+   * @param newStorage the metric storage to use or discard.
+   * @return the {@code newStorage} if no compatible metric is already registered, otherwise the
+   *     previously registered storage.
    */
   @SuppressWarnings("unchecked")
-  public <I extends MetricStorage> I register(I storage) {
-    MetricDescriptor descriptor = storage.getMetricDescriptor();
-    MetricStorage oldOrNewStorage =
-        registry.computeIfAbsent(descriptor.getName().toLowerCase(), key -> storage);
-    // Metric didn't already exist in registry, return.
-    if (storage == oldOrNewStorage) {
-      return (I) oldOrNewStorage;
+  public <I extends MetricStorage> I register(I newStorage) {
+    MetricDescriptor descriptor = newStorage.getMetricDescriptor();
+    I oldOrNewStorage;
+    List<MetricStorage> storages;
+    synchronized (lock) {
+      oldOrNewStorage = (I) registry.computeIfAbsent(descriptor, key -> newStorage);
+      // If storage was NOT added to the registry, its description was a perfect match to one
+      // previously registered and we can skip detecting identity conflicts
+      if (newStorage != oldOrNewStorage || !logger.isLoggable(Level.WARNING)) {
+        return oldOrNewStorage;
+      }
+      storages = new ArrayList<>(registry.values());
     }
-    // Make sure the storage is compatible.
-    if (!oldOrNewStorage.getMetricDescriptor().isCompatibleWith(descriptor)) {
-      throw new DuplicateMetricStorageException(
-          oldOrNewStorage.getMetricDescriptor(),
-          descriptor,
-          "Metric with same name and different descriptor already created.");
+    // Else, we need to look for identity conflicts with previously registered storages
+    for (MetricStorage storage : storages) {
+      // Skip the newly registered storage
+      if (storage == newStorage) {
+        continue;
+      }
+      MetricDescriptor existing = storage.getMetricDescriptor();
+      // Check compatibility of metrics which share the same case-insensitive name
+      if (existing.getName().equalsIgnoreCase(descriptor.getName())
+          && !existing.isCompatibleWith(descriptor)) {
+        logger.log(Level.WARNING, DebugUtils.duplicateMetricErrorMessage(existing, descriptor));
```

Comment:
Let's use the `Supplier` version of `logging.log`

Reply:
Android doesn't support the supplier version. We do exit early if `WARNING` isn't loggable.